framework/src/modules/chain/modules/blocks/verify.js

```diff
@@ -71,71 +72,50 @@ class Verify {
  * FIXME: Some checks are probably redundant, see: logic.transactionPool
  *
  * @private
- * @func checkTransaction
+ * @func checkTransactions
  * @param {Object} block - Block object
  * @param {Object} transaction - Transaction object
  * @param  {boolean} checkExists - Check if transaction already exists in database
  * @param {function} cb - Callback function
  * @returns {function} cb - Callback function from params (through setImmediate)
  * @returns {Object} cb.err - Error if occurred
  */
-__private.checkTransaction = function(block, transaction, checkExists, cb) {
-	async.waterfall(
-		[
-			function getTransactionId(waterCb) {
-				try {
-					// Calculate transaction ID
-					// FIXME: Can have poor performance, because of hash calculation
-					transaction.id = library.logic.transaction.getId(transaction);
-				} catch (e) {
-					return setImmediate(waterCb, e.toString());
-				}
-				// Apply block ID to transaction
-				transaction.blockId = block.id;
-				return setImmediate(waterCb);
-			},
-			function getAccount(waterCb) {
-				// Get account from database if any (otherwise cold wallet)
-				// DATABASE: read only
-				modules.accounts.getAccount(
-					{ publicKey: transaction.senderPublicKey },
-					waterCb
-				);
-			},
-			function verifyTransaction(sender, waterCb) {
-				// Check if transaction id valid against database state (mem_* tables)
-				// DATABASE: read only
-				library.logic.transaction.verify(
-					transaction,
-					sender,
-					null,
-					checkExists,
-					waterCb,
-					null
-				);
-			},
-		],
-		waterCbErr => {
-			if (waterCbErr && waterCbErr.match(/Transaction is already confirmed/)) {
-				// Fork: Transaction already confirmed.
-				modules.delegates.fork(block, 2);
-				// Undo the offending transaction.
-				// DATABASE: write
-				return modules.transactions.undoUnconfirmed(
-					transaction,
-					undoUnconfirmedErr => {
-						modules.transactions.removeUnconfirmedTransaction(transaction.id);
-						return setImmediate(cb, undoUnconfirmedErr || waterCbErr);
-					}
-				);
+__private.checkTransactions = async (transactions, checkExists) => {
+	if (transactions.length === 0) {
+		return;
+	}
+
+	if (checkExists) {
+		const persistedTransactions = await library.storage.entities.Transaction.get(
+			{
+				id_in: transactions.map(transaction => transaction.id),
 			}
-			return setImmediate(cb, waterCbErr);
+		);
+
+		if (persistedTransactions.length > 0) {
+			// Only returning the error of the first transaction to keep the same interface and functionality
+			modules.transactions.removeUnconfirmedTransaction(
+				persistedTransactions[0].id
+			);
+			throw `Transaction is already confirmed: ${persistedTransactions[0].id}`;
 		}
+	}
+
+	const transactionResponses = await modules.processTransactions.verifyTransactions(
+		transactions
 	);
+
+	const unverifiableTransactionResponse = transactionResponses.find(
```

Comment:
shouldn't this be filter? also, `unverifiedTransactionResponse`?

Reply:
This is not a filter because I am keeping the same functionality as the existing implementation.